homeassistant/components/tuya/__init__.py

```diff
@@ -107,86 +123,144 @@ def load_devices(device_list):
+                for component in hass.data[DOMAIN][ENTRY_IS_SETUP]
+            ]
+        )
+    )
+    if unload_ok:
+        hass.data[DOMAIN][ENTRY_IS_SETUP] = set()
+        hass.data[DOMAIN][TUYA_TRACKER]()
+        hass.data[DOMAIN][TUYA_TRACKER] = None
+        hass.data[DOMAIN][TUYA_DATA] = None
+        hass.services.async_remove(DOMAIN, SERVICE_FORCE_UPDATE)
+        hass.services.async_remove(DOMAIN, SERVICE_PULL_DEVICES)
+        hass.data.pop(DOMAIN)
+
+    return unload_ok
+
+
 class TuyaDevice(Entity):
     """Tuya base device."""
 
-    def __init__(self, tuya):
+    def __init__(self, tuya, platform):
         """Init Tuya devices."""
-        self.tuya = tuya
+        self._tuya = tuya
+        self._platform = platform
 
     async def async_added_to_hass(self):
         """Call when entity is added to hass."""
-        dev_id = self.tuya.object_id()
+        dev_id = self._tuya.object_id()
         self.hass.data[DOMAIN]["entities"][dev_id] = self.entity_id
         async_dispatcher_connect(self.hass, SIGNAL_DELETE_ENTITY, self._delete_callback)
         async_dispatcher_connect(self.hass, SIGNAL_UPDATE_ENTITY, self._update_callback)
 
     @property
     def object_id(self):
         """Return Tuya device id."""
-        return self.tuya.object_id()
+        return self._tuya.object_id()
 
     @property
     def unique_id(self):
         """Return a unique ID."""
-        return f"tuya.{self.tuya.object_id()}"
+        return f"tuya.{self._tuya.object_id()}"
 
     @property
     def name(self):
         """Return Tuya device name."""
-        return self.tuya.name()
+        return self._tuya.name()
 
     @property
     def available(self):
         """Return if the device is available."""
-        return self.tuya.available()
+        return self._tuya.available()
+
+    @property
+    def device_info(self):
+        """Return a device description for device registry."""
+        _device_info = {
+            "identifiers": {(DOMAIN, f"{self.unique_id}")},
+            "manufacturer": TUYA_PLATFORMS.get(self._platform, self._platform),
+            "name": self.name,
+            "model": self._tuya.object_type(),
+        }
+        return _device_info
 
     def update(self):
         """Refresh Tuya device data."""
-        self.tuya.update()
+        self._tuya.update()
 
     @callback
-    def _delete_callback(self, dev_id):
+    async def _delete_callback(self, dev_id):
         """Remove this entity."""
         if dev_id == self.object_id:
-            self.hass.async_create_task(self.async_remove())
+            entity_registry = (
+                await self.hass.helpers.entity_registry.async_get_registry()
+            )
+            if entity_registry.async_is_registered(self.entity_id):
+                entity_registry.async_remove(self.entity_id)
+            else:
+                await self.async_remove()
 
     @callback
-    def _update_callback(self):
+    async def _update_callback(self):
```

Comment:
```suggestion
    def _update_callback(self):
```

No need to add the `async` since nothing is being awaited and its already marked as a `@callback`

Reply:
May be I should "throttle" the update method to avoid issue with multiple call to the update service?

Reply:
I'm not sure what the service limits are, but lets save that for a followup PR if its needed as this one is already a bit large.